batch/batch/worker/worker.py

```diff
@@ -478,19 +478,30 @@ def __init__(self, job, name, spec, client_session: httpx.ClientSession, worker:
     async def run(self):
         try:
             async def localize_rootfs():
-                async with image_lock.reader_lock:
-                    # FIXME Authentication is entangled with pulling images. We need a way to test
-                    # that a user has access to a cached image without pulling.
-                    await self.pull_image()
-                    self.image_config = image_configs[self.image_ref_str]
-                    self.image_id = self.image_config['Id'].split(":")[1]
-                    self.worker.image_data[self.image_id] += 1
-
-                    self.rootfs_path = f'/host/rootfs/{self.image_id}'
-                    async with self.worker.image_data[self.image_id].lock:
-                        if not os.path.exists(self.rootfs_path):
-                            await asyncio.shield(self.extract_rootfs())
-                            log.info(f'Added expanded image to cache: {self.image_ref_str}, ID: {self.image_id}')
+                async def _localize_rootfs():
+                    async with image_lock.reader_lock:
+                        # FIXME Authentication is entangled with pulling images. We need a way to test
+                        # that a user has access to a cached image without pulling.
+                        await self.pull_image()
+                        self.image_config = image_configs[self.image_ref_str]
+                        self.image_id = self.image_config['Id'].split(":")[1]
+                        self.worker.image_data[self.image_id] += 1
+
+                        self.rootfs_path = f'/host/rootfs/{self.image_id}'
+
+                        image_data = self.worker.image_data[self.image_id]
+                        async with image_data.lock:
+                            if not image_data.extracted:
+                                try:
+                                    await self.extract_rootfs()
+                                    image_data.extracted = True
+                                    log.info(f'Added expanded image to cache: {self.image_ref_str}, ID: {self.image_id}')
+                                except asyncio.CancelledError:
+                                    raise
+                                except Exception:
+                                    log.exception(f'while extracting image {self.image_ref_str}, ID: {self.image_id}')
+                                    await check_shell(f'rm -Rf {self.rootfs_path}')
```

Comment:
I appreciate this making the extraction idempotent. Why not do this in the case of the cancelled error as well?

Reply:
My understanding is that the CancelledError was for when the application is closing and not us cancelling the task since we shield it. If we catch all exceptions including that one, then we'll try and delete files upon exiting. I could be wrong though!

Reply:
Ah I see, I was forgetting that the CancelledError could come from application shutdown. Thanks!

Reply:
One last thing: can we use `await blocking_to_async(self.pool, shutil.rmtree, image_path)` instead of shelling out to `rm`?